fix(estree/ast): fix `IdentifierName` and `IdentifierReference`

hi-ogawa opened the pull request:
Binding side is tricky https://github.com/oxc-project/oxc/pull/9822, so I went with chipping away these two first.

hi-ogawa commented:
* **#9863** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9863?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/9863?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=9863).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

hi-ogawa commented on crates/oxc_ast/src/serialize.rs:
```diff
@@ -538,6 +571,9 @@ impl ESTree for ClassImplements<'_, '_> {
     raw_deser = "
         const ident = DESER[Box<IdentifierReference>](POS);
         ident.type = 'JSXIdentifier';
+        delete ident.decorators;
+        delete ident.optional;
+        delete ident.typeAnnotation;
```
I went with this to counter against `IdentifierReference` raw deser for `JSXElementIdentifierReference`.

overlookmotel replied:
I'll try and make that a bit faster in a follow-up.

hi-ogawa committed:
fix(estree/ast): fix `IdentifierName` and `IdentifierReference` (#9863)

Binding side is tricky https://github.com/oxc-project/oxc/pull/9822, so I went with chipping away these two first.
